[INFRA-2252][INFRA-910][WIP] - code signing settings for Jenkins Core release automation

olblak opened the pull request:
See [INFRA-2252](https://issues.jenkins-ci.org/browse/INFRA-2252).

This PR is currently being tested on [release.ci.jenkins.io](https://release.ci.jenkins.io) which required [vpn](https://github.com/jenkins-infra/openvpn#howto-get-client-access) access
That jenkins configuration instance is located on [jenkins-infra/charts](https://github.com/jenkins-infra/charts/blob/master/helmfile.d/release.yaml) repository 

ToDo:
[ ] Move kubernetes agent configuration to the Jenkins configuration located [here](https://github.com/jenkins-infra/charts/blob/26e843566331785586b02aed87e26ba28ff51dce/config/release/jenkins/values.yaml#L5)
[ ] Retrieve the GPG key from Azure Key Vault instead of an Azure blob storage

### Proposed changelog entries

* Internal: Add a bash script and a Jenkinsfile to build Jenkins core releases
* Internal: Update maven-deploy-plugin

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

olblak committed:
Update maven-deploy-plugin version && release profile

olblak committed:
Add a Jenkinsfile for releasing jenkins core

olblak committed:
Add *.pem files to gitignore

oleg-nenashev commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+      kubernetes {
+        label 'jenkins-release'
+        defaultContainer 'jnlp'
+        yaml '''
```
Isn't it an obsolete way to define settings in the plugin?

olblak replied:
Not according the [documentation ](https://github.com/jenkinsci/kubernetes-plugin#declarative-pipeline) but to me it would make more sense to move container configuration to jenkins instance configuration instead of the Jenkinsfile which is what I mean when I say 'Todo: Move kubernetes agent configuration to the Jenkins configuration located here'

jglick replied:
`yamlFile` is nice for syntax highlighting at least.

olblak replied:
@jglick Sorry I missed your comment here, and yes I need credentials as I want to be able to trigger a release based  on private repository like jenkinsci-cert. I am going to close this PR as I created a new repository that contains, so I will probably implement your suggestion [here](https://github.com/jenkins-infra/release/blob/master/Jenkinsfile.d/release)

oleg-nenashev reviewed (commented):
I am not 100% sure it should be a part of the core repo. We might need to spin releases for older core and LTS baselines with newer scripts, and in such case checking out the Jenkins core repo twice cause a serious risk of mistakes. It worth a discussion IMHO

olblak commented:
> I am not 100% sure it should be a part of the core repo. We might need to spin releases for older core and LTS baselines with newer scripts, and in such case checking out the Jenkins core repo twice cause a serious risk of mistakes. It worth a discussion IMHO

@oleg-nenashev  It's a good point, I don't have a strong opinion neither, when I started working on this, I hesitated to create this PR either on this repository or jenkinsci/packaging.

The reason why I put this in this repository was because I wanted to trigger a new release each time master branch was updated.
But then I created a different Jenkinsfile for packagings steps on jenkinsci/packaging with a work in progress PR located [here](https://github.com/jenkinsci/packaging/pull/120)

While I think it's better to have two different job one for releasing and a second one for packaging, I could definitely move Jenkinsfile.release (and build script) on jenkinsci/packaging

oleg-nenashev commented:
> While I think it's better to have two different job one for releasing and a second one for packaging, I could definitely move Jenkinsfile.release (and build script) on jenkinsci/packaging

I do not have strong opinion here, the current implementation may work as well. Let;s ask @jenkinsci/core to get more feedback before acting on it.

jglick commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+    workdir: /home/jenkins
+    command:
+    - cat
+    tty: true
```
BTW I recommend the more legible

```yaml
command:
- sleep
args:
- 99d
```

(`infinity` works in most, but not all, versions of `sleep`.)

olblak replied:
thx for the suggestion

jglick replied:
To be clear, that substitutes for

```yaml
    command:
    - cat
    tty: true
```

which has the same effect but in a more mysterious way.

jglick commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+    agent {
+      kubernetes {
+        label 'jenkins-release'
+        defaultContainer 'jnlp'
```
```suggestion
```

Unnecessary I think.

olblak replied:
You were right and I delete from [here](https://github.com/jenkins-infra/release/blob/master/Jenkinsfile.d/release#L14-L11)

jglick commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+    }
+    triggers {
+      // Trigger a new release once a week on Sunday
+      cron('H H(0-7) * * 0')
```
I think for a job of this importance, we can set a specific time of day, not rely on `H`!

olblak replied:
As a first iteration,I just remove this cron as  we'll trigger the release manually

jglick commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+
+        steps {
+          container('azure-cli') {
+            // TODO move into a shell script in the repository
```
yes please

olblak replied:
done

olblak replied:
[here](https://github.com/jenkins-infra/release/blob/master/utils/release.sh)

jglick commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+              script {
+                def properties = readProperties file: 'release.properties'
+                env.RELEASE_SCM_TAG = properties['scm.tag']
+                env.RELEASE_VERSION = properties['project.rel.org.jenkins-ci.main:jenkins-war']
```
Seems to be unused?

jglick commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+              '''
+              script {
+                def properties = readProperties file: 'release.properties'
+                env.RELEASE_SCM_TAG = properties['scm.tag']
```
If you merge with the following stage, the logic could all be moved into an external (e.g., shell) script without using the `readProperties` step or a `script` block.

jglick commented on pom.xml:
```diff
@@ -682,6 +682,31 @@ THE SOFTWARE.
+                  <!-- Skipp Test <goals>-DskipTests -Danimal.sniffer.skip=false javadoc:javadoc deploy</goals>-->
+              <goals>-Danimal.sniffer.skip=false javadoc:javadoc deploy</goals>
+              <!--<goals>javadoc:javadoc deploy</goals>-->
+              <pushChanges>false</pushChanges>
```
How I have come to loathe `maven-release-plugin`…

jeffret-b replied:
I got to that point years ago and the feeling hasn't gotten any better.

jglick commented on scripts/buildJenkins.bash:
```diff
@@ -0,0 +1,172 @@
+#!/bin/bash
+
+set -e    # Abort script at first error
+set -u    # Attempt to use undefined variable outputs error message, and forces an exit
```
Suggest:

```bash
set -euxo pipefail
```

olblak replied:
[done](https://github.com/jenkins-infra/release/blob/96588a42488b165e3e0c04dd1fed422a5a38297a/utils/release.sh#L3)

jglick reviewed (commented):
As to where this should live, I would ask simply: Is there any plausible circumstance under which someone doing code modifications to Jenkins core would wish to touch any of these release files? I suspect not—they all seem rather generic and focused on infrastructure rather than the specific content of the project (beyond it using Maven). That would _suggest_ moving it to somewhere on `jenkins-infra` and not attempting to keep it synchronized to the content of this repository except under rare circumstances (such as Jenkins switching to JDK 11 at build time).

oleg-nenashev reviewed (changes requested):
I am requesting changes because of a custom Docker image being used. After looking deeper into the code, I also tend to recommend keeping all bits (Jenkinsfile, Build/Release scripts and Docker images) in a separate repository inside jenkins-infra. It will simplify management and also let us to add specific test automation and CI later

olblak committed:
Move release script in a different git repository called jenkins-infra/release

olblak commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+spec:
+  containers:
+  - name: maven
+    image: olblak/maven:3-jdk-8
```
I totally agree, test fails If I run maven as the root user but I disabled the tests in order to switch back to 'maven:3-jdk8' as defined [here](https://github.com/jenkins-infra/release/blob/1c2409392ddae2d744c1f6b7146ed7cdc8148c7e/PodTemplates.d/release-linux.yaml#L34)

olblak commented on Jenkinsfile.release:
```diff
@@ -0,0 +1,206 @@
+    * Azure-Credentials
+    * SSH-agent
+
+  docker image olblak/maven is just a customized docker image based on maven where maven is run as
```
I switch back to the official maven image, as I am skipping the test anyway

olblak replied:
[here](https://github.com/jenkins-infra/release/blob/96588a42488b165e3e0c04dd1fed422a5a38297a/PodTemplates.d/release-linux.yaml#L34)

olblak commented:
I split the content of this PR in a different repository located [here](https://github.com/jenkins-infra/release).

This new repository contains Jenkinsfile, podTemplates, and scripts needed to release and package jenkins-core.

I am keeping this PR open right now as I still have to modify the pom.xml.
Instead of touching this file, I tried to configured everything at the cli level from the script located [here](https://github.com/jenkins-infra/release/blob/1c2409392ddae2d744c1f6b7146ed7cdc8148c7e/utils/release.sh#L207) but for some reasons it didn't work as expected.

oleg-nenashev commented on pom.xml:
```diff
@@ -610,6 +635,24 @@ THE SOFTWARE.
+          </plugin>
+          <plugin>
+            <artifactId>maven-jarsigner-plugin</artifactId>
+            <configuration>
```
Such implementation will not work on my machine with a hardware codesigning key. But I;d guess it will be fine for @kohsuke, so why not.

I suggest adding some flags to enforce the code signing verification, e.g. sample from Remoting

```
<configuration>
                  <processMainArtifact>true</processMainArtifact>
                  <processAttachedArtifacts>false</processAttachedArtifacts>
                  <certs>true</certs>
                  <errorWhenNotSigned>true</errorWhenNotSigned>
                  <arguments>-strict</arguments> <!--otherwise certificate chains will be ignored-->
                  <alias>${hudson.sign.alias}</alias>
                  <storepass>${hudson.sign.storepass}</storepass>
                  <keystore>${hudson.sign.keystore}</keystore>
                  <storetype>${hudson.sign.storetype}</storetype>
                  <providerClass>${hudson.sign.providerClass}</providerClass>
                  <providerArg>${hudson.sign.providerArg}</providerArg>
                </configuration>
```

oleg-nenashev replied:
```suggestion
            <configuration>
              <certs>true</certs>
              <errorWhenNotSigned>true</errorWhenNotSigned>
              <arguments>-strict</arguments> <!--otherwise certificate chains will be ignored-->
```

jeffret-b replied:
I think the goal is to be able to use a software signing certificate, obtained on behalf of the CDF. And for Kohsuke to also be able to use it, at least for the interim till we get everything working correctly. If it doesn't work for a hardware-based codesigning key, that's probably not a concern.

jeffret-b replied:
+1 on enforcing the code signing when this profile is used.

olblak replied:
https://github.com/jenkins-infra/release/blob/master/utils/release.sh#L221-L219

oleg-nenashev reviewed (changes requested):
In remoting code signing I had multiple issues with code signing being broken due to the signing issues (usually broken store setup). I suggest to explicitly add certificate validation so that we prevent releases from broken certificates from being shipped

timja commented:
> I split the content of this PR in a different repository located [here](https://github.com/jenkins-infra/release).
> 
> This new repository contains Jenkinsfile, podTemplates, and scripts needed to release and package jenkins-core.
> 
> I am keeping this PR open right now as I still have to modify the pom.xml.
> Instead of touching this file, I tried to configured everything at the cli level from the script located [here](https://github.com/jenkins-infra/release/blob/1c2409392ddae2d744c1f6b7146ed7cdc8148c7e/utils/release.sh#L207) but for some reasons it didn't work as expected.

In case you're interested,  I've added azure key vault support to declarative pipeline in the latest release (https://github.com/jenkinsci/azure-keyvault-plugin/releases) it might simplify your release script, especially for retrieving certificates.

olblak commented:
> > I split the content of this PR in a different repository located [here](https://github.com/jenkins-infra/release).
> > This new repository contains Jenkinsfile, podTemplates, and scripts needed to release and package jenkins-core.
> 
> @olblak In case you're interested, I've added support for declarative pipeline in the latest release of azure key vault (https://github.com/jenkinsci/azure-keyvault-plugin/releases) it might simplify your release script, especially for retrieving certificates.
Awesome, for now I have something that works on that parts, so I'll leave as is for the moment, but I may improve it in a next iteration, thanks for the suggestion

jeffret-b reviewed (commented):
This looks like it's heading the right direction and making sense.

Do we want to use the `tsa` option for jarsigner? Seems like that would be a good idea. See https://docs.oracle.com/javase/6/docs/technotes/tools/windows/jarsigner.html#tsa , https://github.com/jenkinsci/jenkins/blob/master/war/pom.xml#L610 , and https://github.com/jenkinsci/remoting/blob/master/pom.xml#L498 .

olblak commented:
I am closing this PR as I moved the current work on [jenkins-infra/release](https://github.com/jenkins-infra/release/blob/master/utils/release.sh#L221-L219), feel free to open issues or PR over there
